Mathlib/RingTheory/MvPowerSeries/Evaluation.lean

```diff
@@ -0,0 +1,408 @@
+  coeToMvPowerSeries_uniformInducing.isDenseInducing coeToMvPowerSeries_denseRange
+
+variable {a : σ → S}
+
+/- The coercion of polynomials into power series is uniformly continuous. -/
+theorem _root_.MvPolynomial.coeToMvPowerSeries_uniformContinuous
+    [UniformAddGroup R] [UniformAddGroup S] [IsLinearTopology S S]
+    (hφ : Continuous φ) (ha : EvalDomain a) :
+    UniformContinuous (MvPolynomial.eval₂Hom φ a) := by
+  classical
+  apply uniformContinuous_of_continuousAt_zero
+  intro u hu
+  simp only [coe_eval₂Hom, (induced_iff_nhds_eq _).mp rfl, coe_zero, mem_map, mem_comap]
+  rw [map_zero, IsLinearTopology.hasBasis_ideal.mem_iff] at hu
+  rcases hu with ⟨I, hI, hI'⟩
+  let tendsto_zero := ha.tendsto_zero
+  let hpow := ha.hpow
+  simp only [tendsto_def] at tendsto_zero hpow
+  specialize tendsto_zero I hI
+  simp only [mem_cofinite] at tendsto_zero
+  let hpow' := fun s ↦ hpow s hI
+  simp only [mem_map, mem_atTop_sets, ge_iff_le, mem_preimage, SetLike.mem_coe] at hpow'
+  let n : σ → ℕ := fun s ↦ sInf {n : ℕ | (a s) ^ n.succ ∈ I}
+  have hn_ne : ∀ s, Set.Nonempty {n : ℕ | (a s) ^ n.succ ∈ I} := fun s ↦ by
+    rcases hpow' s with ⟨n, hn⟩
+    use n
+    simp only [mem_setOf_eq, hn n.succ (Nat.le_succ n)]
+  have hn : Set.Finite (n.support) := by
+    apply @Finite.Set.subset  _ _ _ tendsto_zero
+    intro s
+    simp only [Function.mem_support, ne_eq, mem_compl_iff, mem_preimage, SetLike.mem_coe, not_not,
+      not_imp_comm, imp_or, n, Nat.sInf_eq_zero, mem_setOf_eq, zero_add, pow_one, imp_self, true_or]
+  let n₀ : σ →₀ ℕ := {
+    toFun := n
+    support := hn.toFinset
+    mem_support_toFun := fun (s : σ) ↦ by simp }
+  let D := Iic n₀
+  have hD : Set.Finite D := finite_Iic _
+  use iInter (fun (d : D) ↦ { p | φ (p d.val) ∈ I})
+  rw [nhds_pi, Filter.mem_pi]
+  constructor
+  · use D, hD
+    use fun d ↦ if d ∈ D then φ ⁻¹' I else univ
+    constructor
+    · intro d
+      split_ifs with hd
+      · exact hφ.continuousAt.preimage_mem_nhds (map_zero φ ▸ hI)
+      · exact univ_mem
+    · intro p
+      simp only [Set.mem_pi, mem_ite_univ_right, mem_preimage, SetLike.mem_coe,
+        iInter_coe_set, mem_iInter]
+      exact fun hp i hi ↦ hp i hi hi
+  · intro p hp
+    simp only [iInter_coe_set, mem_preimage, coeToMvPowerSeries.ringHom_apply,
+      mem_iInter, mem_setOf_eq] at hp
+    simp only [mem_preimage]
+    apply hI'
+    simp only [coe_eval₂Hom, SetLike.mem_coe]
+    rw [eval₂_eq]
+    apply Ideal.sum_mem
+    intro d _
+    by_cases hd : d ∈ D
+    · exact Ideal.mul_mem_right _ _ (hp d hd)
+    · apply Ideal.mul_mem_left
+      simp only [mem_Iic, D, Finsupp.le_iff] at hd
+      push_neg at hd
+      rcases hd with ⟨s, hs', hs⟩
+      rw [Finset.prod_eq_prod_diff_singleton_mul hs']
+      · apply Ideal.mul_mem_left
+        rw [← Nat.add_sub_of_le (Nat.succ_le_of_lt hs), pow_add]
+        apply Ideal.mul_mem_right
+        simp only [Finsupp.coe_mk, n₀, n]
+        exact Nat.sInf_mem (hn_ne s)
```

Comment:
I have other comments and suggestions in preparation for this proof, but this is an obvious one: could you extract this to a lemma (or find an existing lemma) saying that a product belongs to an ideal if one of the factors does?